platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -0,0 +1,283 @@
+            @Override
+            protected boolean doCompaction() throws IOException, InterruptedException {
+                return pathToKeyValue.compact(
+                        statisticsUpdater::setLeavesStoreCompactionTimeMs,
+                        statisticsUpdater::setLeavesStoreCompactionSavedSpaceMb);
+            }
+        });
+    }
+
+    /**
+     * Enables background compaction.
+     */
+    void enableBackgroundCompaction() {
+        compactionEnabled.set(true);
+    }
+
+    /**
+     * Stops all compactions in progress and disables background compaction.
+     * All subsequent calls to compacting methods will be ignored until {@link #enableBackgroundCompaction()} is called.
+     */
+    void stopAndDisableBackgroundCompaction() {
+        synchronized (createOrGetCompactingExecutor()) {
+            compactionFuturesByName.forEach((k, v) -> v.cancel());
+            compactionFuturesByName.clear();
+        }
+        compactionEnabled.set(false);
+    }
+
+    /**
+     * Submits a compaction task for execution. If a compaction task for the same storage type is already in progress,
+     * the call is effectively no op.
+     * @param task a compaction task to execute
+     */
+    private void submitCompactionTaskForExecution(CompactionTask task) {
+        if (!compactionEnabled.get()) {
+            return;
+        }
+
+        ExecutorService executor = createOrGetCompactingExecutor();
+
+        synchronized (executor) {
+            if (compactionFuturesByName.containsKey(task.id)) {
+                CompletableFuture<?> completableFuture =
+                        compactionFuturesByName.get(task.id).asCompletableFuture();
+                if (completableFuture.isDone()) {
+                    compactionFuturesByName.remove(task.id);
+                } else {
+                    logger.debug(MERKLE_DB.getMarker(), "Compaction for {} is already in progress", task.id);
+                    return;
+                }
+            }
+
+            InterruptibleCompletableFuture<Boolean> future =
+                    InterruptibleCompletableFuture.runAsyncInterruptibly(task, executor);
+            future.asCompletableFuture().thenAccept(result -> {
+                if (result) {
+                    statisticsUpdater.updateStoreFileStats();
+                    statisticsUpdater.updateOffHeapStats();
+                    logger.info("Finished compaction for " + task.id);
+                }
+            });
+            compactionFuturesByName.put(task.id, future);
+        }
+    }
+
+    /**
+     * Creates a thread pool for compaction tasks if it does not exist yet, or returns the existing one.
+     * If the existing thread pool is shutdown, it will be replaced with a new one. Note that there is only one
+     * compacting thread-pool per JVM.
+     * @return a thread pool for compaction tasks
+     */
+    ExecutorService createOrGetCompactingExecutor() {
```

Comment:
Yes, the real concern is that it can be shutdown (even though we never explicitly do it in the code). If you think that I overthought it, and we shouldn't really have any safeguards against it, I can remove it. We're not supposed to shut down this executor in any case.

Reply:
I don't know of any path that shuts down an executor unless `shutdown` is called.
I think we're better off just creating it once, possibly when the `MerkleDbCompactionCoordinator` is created; particularly if we intend for the lifecycle to be more-or-less the entire life of the platform.

Reply:
I simplified the compaction executor initialization.